pkg/cmd/secret/set/set.go

```diff
@@ -0,0 +1,216 @@
+	cmd.Flags().Lookup("org").NoOptDefVal = "@owner"
+	cmd.Flags().StringVarP(&opts.Visibility, "visibility", "v", "private", "Set visibility for an organization secret: `all`, `private`, or `selected`")
+	cmd.Flags().StringSliceVarP(&opts.RepositoryNames, "repos", "r", []string{}, "List of repository names for `selected` visibility")
+	cmd.Flags().StringVarP(&opts.Body, "body", "b", "-", "Provide either a literal string or a file path; prepend file paths with an @. Reads from STDIN if not provided.")
+
+	return cmd
+}
+
+func setRun(opts *SetOptions) error {
+	body, err := getBody(opts)
+	if err != nil {
+		return fmt.Errorf("did not understand secret body: %w", err)
+	}
+
+	c, err := opts.HttpClient()
+	if err != nil {
+		return fmt.Errorf("could not create http client: %w", err)
+	}
+	client := api.NewClientFromHTTP(c)
+
+	var baseRepo ghrepo.Interface
+	if opts.OrgName == "" || opts.OrgName == "@owner" {
+		baseRepo, err = opts.BaseRepo()
+		if err != nil {
+			return fmt.Errorf("could not determine base repo: %w", err)
+		}
+	}
+
+	host := ghinstance.OverridableDefault()
+	if opts.OrgName == "@owner" {
+		opts.OrgName = baseRepo.RepoOwner()
+		host = baseRepo.RepoHost()
+	}
+
+	var pk *PubKey
+	if opts.OrgName != "" {
+		pk, err = getOrgPublicKey(client, host, opts.OrgName)
+	} else {
+		pk, err = getRepoPubKey(client, baseRepo)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to fetch public key: %w", err)
+	}
+
+	eBody, err := box.SealAnonymous(nil, body, &pk.Raw, opts.RandomOverride)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt body: %w", err)
+	}
+
+	encoded := base64.StdEncoding.EncodeToString(eBody)
+
+	if opts.OrgName != "" {
+		err = putOrgSecret(client, pk, host, *opts, encoded)
+	} else {
+		err = putRepoSecret(client, pk, baseRepo, opts.SecretName, encoded)
+	}
+	if err != nil {
+		return fmt.Errorf("failed to set secret: %w", err)
+	}
+
+	if opts.IO.IsStdoutTTY() {
+		cs := opts.IO.ColorScheme()
+		fmt.Fprintf(opts.IO.Out, "%s Set secret %s\n", cs.SuccessIcon(), opts.SecretName)
+	}
+
+	return nil
+}
+
+func validSecretName(name string) error {
+	if name == "" {
+		return errors.New("secret name cannot be blank")
+	}
+
+	if strings.HasPrefix(name, "GITHUB_") {
+		return errors.New("secret name cannot begin with GITHUB_")
+	}
+
+	leadingNumber := regexp.MustCompile(`^[0-9]`)
+	if leadingNumber.MatchString(name) {
+		return errors.New("secret name cannot start with a number")
+	}
+
+	validChars := regexp.MustCompile(`^([0-9]|[a-z]|[A-Z]|_)+$`)
+	if !validChars.MatchString(name) {
+		return errors.New("secret name can only contain letters, numbers, and _")
+	}
+
+	return nil
+}
+
+func getBody(opts *SetOptions) (body []byte, err error) {
+	if opts.Body == "-" {
+		body, err = ioutil.ReadAll(opts.IO.In)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read from STDIN: %w", err)
+		}
+
+		return
+	}
+
+	if strings.HasPrefix(opts.Body, "@") {
```

Comment:
Would it be possible to set a secret value that starts with a literal `@`, e.g.
```
gh secret add USERNAME -b "@vilmibm"
```

Reply:
no, good point; I was just mimicking what I saw for `gh api`. I'm wondering if it just makes better sense to not magically support filenames here but instead require `cat file | gh secret set foo_secret` ?

Reply:
There is sometimes value in supporting "magic" syntaxes, like the values starting with `@` (I've first observed this pattern in `curl`), but there also needs to be a version of the flag that takes all values literally. In `gh api`, the `-f` flag is the version of `-F` that has no magic processing. So if `-b` supported magic syntax, there should maybe be another version of the flag without the magic.

In this case, however, I think that the `-` filename input would be the most common to combine with piping, and most shells provide very convenient syntax to pipe in input from a filename, that maybe we don't have to include syntax to read from a file at all. 👍